### PR TITLE
Expand 'hikes' endpoint to allow filtering criteria

### PR DESCRIPTION
**Pull request by brandiseeley**
## The endpoint
The 'hikes' endpoint now accepts a JSON body containing search criteria.

- If there isn't any filtration criteria, we return all hikes
- If there is filtration criteria
    - If the attributes are attributes that exist in our hikes, we will filter and return a list of matching hikes, even if there are none
    - If there are any attributes that *don't* exist as attributes of our hikes, we return a `400 Bad Request` response.

## The documentation
As a sort of running documentation for our API, I've created a file called `sample_requests.http` with demo requests. If you haven't already, you can install the "REST client" extension for VSCode. Once you've done this, there will be a `Send Request` button above each request within the `sample_requests.http` file.

### CODE REVIEW

**Review by sayan-gg — APPROVED** [review]
The `sample_requests.http` document + REST Client extension is cool

Nice that we can start filtering hikes

**Comment by brandiseeley** [comment]
I just added more to this. Do you mind checking it out before I merge it? @sayan-gg 

Now, instead of a list of attribute, we need to send a list of filters. Here's an example:

```
GET http://localhost:8000/gorapass/hikes HTTP/1.1
content-type: application/json

{
    "filters": [
        {
            "attribute_name": "total_elevation_gain",
            "attribute_value": 3000,
            "filter_type": "less_than"
        },
        {
            "attribute_name": "total_elevation_gain",
            "attribute_value": 2000,
            "filter_type": "greater_than"
        }
    ]
}
```

We can add as many filters as we'd like. All filters need to have an `attribute_name`, `attribute_value`, and `filter_type`.

File: gorapass/views.py [hunk 1]
```diff
@@ -23,9 +35,49 @@ def hike(request, hike_id):
...
+            case 'exact':
+                hike_models = [ hike for hike in hike_models if getattr(hike, attribute) == value ]
+            case 'partial':
+                hike_models = [ hike for hike in hike_models if value.lower() in str(getattr(hike, attribute)).lower() ]
+            case 'less_than':
+                hike_models = [ hike for hike in hike_models if getattr(hike, attribute) < value ]
+            case 'greater_than':
+                hike_models = [ hike for hike in hike_models if getattr(hike, attribute) > value ]
+                pass
+            case _:
```
Comment (sayan-gg) [inline gorapass/views.py:73]: What is `case _:` used for?

File: gorapass/views.py [hunk 2]
```diff
@@ -23,9 +35,49 @@ def hike(request, hike_id):
...
+        if not hasattr(Hikes, filter[ATTRIBUTE_NAME]):
+            return False
+        
+    return True
+
+def filter_hikes(hike_models, filters):
+    for filter in filters:
+        attribute = filter[ATTRIBUTE_NAME]
+        value = filter[ATTRIBUTE_VALUE]
+        match filter[FILTER_TYPE]:
```
Comment (sayan-gg) [inline gorapass/views.py:63]: I haven't seen the `match` statement used before. This is cool. So basically, you are whittling down the list of hike_models with every loop to filter it down.

File: gorapass/views.py [hunk 3]
```diff
@@ -23,9 +35,49 @@ def hike(request, hike_id):
...
+            case 'less_than':
+                hike_models = [ hike for hike in hike_models if getattr(hike, attribute) < value ]
+            case 'greater_than':
+                hike_models = [ hike for hike in hike_models if getattr(hike, attribute) > value ]
+                pass
+            case _:
+                pass
+        # Return early if we run out of matching hikes
+        if len(hike_models) == 0:
+            return hike_models
```
Comment (sayan-gg) [inline gorapass/views.py:77]: This is a nice added escape hatch to improve performance 👍

File: gorapass/views.py [hunk 4]
```diff
@@ -23,9 +35,49 @@ def hike(request, hike_id):
...
+        hike_models = filter_hikes(hike_models, filters)
+
     hike_dicts = [ model_to_dict(hike) for hike in hike_models ]
     return JsonResponse(hike_dicts, safe=False)
 
+def valid_hike_filters(filters):
+    for filter in filters:
+        if set(filter.keys()) != FILTER_KEYS:
+            return False
+        if not hasattr(Hikes, filter[ATTRIBUTE_NAME]):
```
Comment (sayan-gg) [inline gorapass/views.py:54]: There's a chance in the future that the returned filters payload will be a JSON with all the different filters listed and just a null value next to the filter that isn't used. I think this is fine for now but we might want to revisit this in the future depending on how we end up setting up the post request for filters

**Review by sayan-gg — APPROVED** [review]
Nice work setting up multiple filter criteria

File: gorapass/views.py (hunk 1 above)
Comment (brandiseeley) [reply gorapass/views.py:73]: Ahh, I forgot about that. In theory, we shouldn't make it to the default case, but if we do we should probably return a 400 bad request or a 500 server error. The validation function should catch any filters that aren't present. 

I think for now we can roll with this so we don't block things, but let's add something like "validate filter types" as a todo.

File: gorapass/views.py (hunk 2 above)
Comment (brandiseeley) [reply gorapass/views.py:63]: Exactly. I think this is better than checking each hike for *all* filters.

File: gorapass/views.py (hunk 3 above)
Comment (brandiseeley) [reply gorapass/views.py:77]: :+1:

File: gorapass/views.py (hunk 4 above)
Comment (brandiseeley) [reply gorapass/views.py:54]: Yeah could be. Things also change if we end up with hikes that have different attributes.

**Comment by brandiseeley** [comment]
@sayan-gg I noted your comment about the default case and will add it to my list, but for now this is working, just need to tighten up validation.

**Pull request merged**